### PR TITLE
feat(models): add StreamPETR building blocks

### PR DESCRIPTION
**Pull request by vividf**
## Summary
Adds the reusable building blocks a PETR-style streaming camera detector needs, with no detector wired up yet. Every file is either new or an additive extension, so nothing existing changes behavior.

## Changes
- **VoVNet-99 multiscale backbone** (`models/common/backbones/vovnet.py`): OSA blocks with eSE attention, `out_features` selection over `stem`/`stage2`–`stage5`, `frozen_stages` and `norm_eval` handling. Parameter names match the published StreamPETR checkpoints, which is what makes pretrained-weight loading exact.
- **GridMask** (`models/common/grid_mask.py`): training-time structured occlusion augmentation.
- **NMS-free 3D box coder** (`task_modules/bbox_coders.py`): `normalize_boxes3d` / `denormalize_boxes3d` define the shared regression space (metric centers, log sizes, split sin/cos yaw) used by both the losses and decoding; `NMSFreeBBoxCoder3D` decodes one-to-one-matched queries by flat top-k over query x class scores, with optional score threshold and `post_center_range` filtering.
- **`BBox3DL1Cost`** (`task_modules/match_costs.py`): pairwise L1 cost in that encoding with per-channel `code_weights`, so velocity channels can be excluded from matching.
- **Temporal-query helpers** (`task_modules/streaming.py`): sine/cosine position embeddings (`pos2posemb3d` concatenates in the (y, x, z) order the pretrained weights expect), NeRF-style encoding for ego motion, `memory_refresh`, `topk_gather`, ego-frame reference-point transforms, and the SE / modulated-LayerNorm blocks.

Two robustness fixes found while writing the tests:
- `out_features` is now validated at construction; a typo previously produced an empty output tuple and failed later inside the neck with a confusing error.
- `_freeze_stages()` now also runs in `__init__`, so `frozen_stages` takes effect before the first `train()` call (it otherwise leaked frozen parameters into optimizer groups built during setup).

## Testing
New unit tests next to the modules:
- `models/common/backbones/tests/test_vovnet.py` — stage shapes/channels, `out_features` selection and validation, `norm_eval`, `frozen_stages`.
- `models/common/tests/test_grid_mask.py` — eval/probability pass-through, masking keeps shape and leaves a grid.
- `task_modules/tests/test_bbox_coders_nms_free.py` — encode/decode round trip, log-size and sin/cos layout, flat top-k index to class/box mapping, score threshold and `post_center_range` filtering, per-sample outputs.
- `task_modules/tests/test_streaming.py` — the (y, x, z) embedding order, 2*pi scaling and sin/cos interleaving, memory refresh, top-k gather, pose application, inverse-sigmoid inversion and clamping.

## Stack
1. feat(datamodule): add scene-streaming dataloaders for temporal detectors — #116
2. fix(transforms): fold ResizeCropFlipRotImage crop into the aug matrix exactly — #117
3. **feat(models): add StreamPETR building blocks ← this PR**
4. feat(models): add the StreamPETR detector — [vividf#9](https://github.com/vividf/autoware-ml/pull/9)
5. feat(streampetr): add base nuScenes and T4 j6gen2 configurations — [vividf#10](https://github.com/vividf/autoware-ml/pull/10)
6. feat(streampetr): harden the streaming training infrastructure — [vividf#3](https://github.com/vividf/autoware-ml/pull/3)
7. feat(streampetr): project 2D annotations and fold augmentations into ego poses — [vividf#4](https://github.com/vividf/autoware-ml/pull/4)
8. feat(streampetr): add auxiliary 2D head, partial-ignore, and DDP loss normalization — [vividf#5](https://github.com/vividf/autoware-ml/pull/5)
9. feat(streampetr): pin the production three-stage training configs — [vividf#6](https://github.com/vividf/autoware-ml/pull/6)

#116, #117 and #118 are independent of each other and all based on `main`, so they can be reviewed and merged in any order. Each later layer is opened here once its prerequisites merge, so every PR stays a single-layer diff.

### CODE REVIEW

**Comment by amadeuszsz** [comment]
@codex review

File: autoware_ml/models/common/grid_mask.py [hunk 1]
```diff
@@ -0,0 +1,108 @@
...
+        self.ratio = ratio
+        self.mode = mode
+        self.prob = prob
+
+    def forward(self, x: torch.Tensor) -> torch.Tensor:
+        """Apply one shared grid mask to a ``(N, C, H, W)`` image batch."""
+        if np.random.rand() > self.prob or not self.training:
+            return x
+        n, c, h, w = x.size()
+        x = x.view(-1, h, w)
```
Comment (amadeuszsz) [inline autoware_ml/models/common/grid_mask.py]: Prefer `reshape` over `view`. [When possible, reshape returns a view of the original tensor.](https://www.geeksforgeeks.org/deep-learning/whats-the-difference-between-reshape-and-view-in-pytorch/)

**Review by amadeuszsz — COMMENTED** [review]
We already have `GridMask` transform (cpu). What is the plan for this class here? If you intend to make it a part of preprocessing pipeline (GPU, after all the transforms), then maybe `models` is not the best namespace for it.

File: autoware_ml/models/common/grid_mask.py [hunk 2]
```diff
@@ -0,0 +1,108 @@
...
+                s = d * i + st_h
+                t = min(s + band, padded_h)
+                mask[s:t, :] *= 0
+        if self.use_w:
+            for i in range(padded_w // d):
+                s = d * i + st_w
+                t = min(s + band, padded_w)
+                mask[:, s:t] *= 0
+
+        r = np.random.randint(self.rotate)
```
[the remark this answers is not in the document]
Comment (vividf) [reply autoware_ml/models/common/grid_mask.py]: Fixed in cd5c5bb: `rotate=0` now disables rotation instead of sampling. Added a test for it.

File: autoware_ml/models/common/backbones/tests/test_vovnet.py [hunk 3]
```diff
@@ -0,0 +1,56 @@
+"""Unit tests for the VoVNet multiscale backbone."""
```
[the remark this answers is not in the document]
Comment (vividf) [reply autoware_ml/models/common/backbones/tests/test_vovnet.py:1]: Fixed in cd5c5bb: CI now runs `pytest autoware_ml/ -v` (same as the pixi `test` task), so module-local tests are collected.

File: autoware_ml/models/common/grid_mask.py (hunk 1 above)
Comment (vividf) [reply autoware_ml/models/common/grid_mask.py]: Done in cd5c5bb, switched to `reshape`.

**Comment by vividf** [comment]
@amadeuszsz 
Agreed, this is a GPU-side batch augmentation, not part of the model. In cd5c5bb I moved it to `autoware_ml/preprocessing/camera/` as a `DataPreprocessing` layer (`__call__(batch, *, is_training)`) and renamed it `BatchGridMask`.

It is the upstream StreamPETR variant kept verbatim so we can reproduce the baseline first. The docstring lists how it differs from the CPU `transforms.camera.masking.GridMask` (one shared mask per batch, period range, mask polarity, rotation semantics), and there is a TODO to consolidate the two once the baseline is reproduced. The detector will pick it up via `data_preprocessing.pipeline` in the follow-up PR.

**Review by amadeuszsz — APPROVED** [review]
LGTM

**Comment by amadeuszsz** [comment]
> @amadeuszsz Agreed, this is a GPU-side batch augmentation, not part of the model. In [cd5c5bb](https://github.com/tier4/autoware-ml/commit/cd5c5bb5e375e7d3c4aa067017d17f68ba4e36f8) I moved it to `autoware_ml/preprocessing/camera/` as a `DataPreprocessing` layer (`__call__(batch, *, is_training)`) and renamed it `BatchGridMask`.

Please be aware that preprocessing pipeline is applied after data augmentation. Worth to know if you need run grid mask before some of transforms.
